### PR TITLE
[Pactus]: Support Pactus Blockchain

### LINKED ISSUE #4056: Add support for Pactus Blockchain
## Description
Pactus is an open source, layer-1 blockchain technology with a revolutionary, secure, Solid State Proof of Stake Consensus.

Name: [Pactus](https://pactus.org)
HD Derivation Scheme: `m/44'/21888'/3'/0'`
Symbol: PAC
Documentation: https://docs.pactus.org
Reason:

## Checklist

Finish check list [here](https://developer.trustwallet.com/wallet-core/newblockchain#blockchain-checklist).

### PR DESCRIPTION
**Pull request by b00f**
## Description

This PR adds support for the [Pactus](https://pactus.org) Blockchain in TrustWalletCore. 
It follows the [Integration Criteria](https://developer.trustwallet.com/developer/wallet-core/newblockchain#integration-criteria) highlighted by TrustWalletCore team.

## How to test

This PR includes all the necessary tests to ensure the integration is done properly. 
We have also compared the results with the native wallet in Pactus, which is available [here](https://github.com/pactus-project/pactus/blob/main/crypto/ed25519/ed25519_test.go) and [here](https://github.com/pactus-project/pactus/tree/main/wallet).

This PR fixes #4056 .

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

## Integration Checklist

- [x] Coin Definition:
  - [x] Add the coin definition to `registry.json`.
  - [x] Execute `pushd codegen-v2 && cargo run -- new-blockchain <coinid> && popd` to generate blockchain skeleton and configure integration tests.
  - [x] Execute `tools/generate-files`.
  - [x] Execute `pushd rust && cargo check --tests && popd` to check if Rust codebase compiles successfully.
  - [x] Execute `make -Cbuild -j12 tests TrezorCryptoTests` to check if C++ codebase compiles successfully.
- [x] Add relevant constants in `Curve`, `Hasher` etc., in C++ and Rust, as necessary.
- [x] Implement functionality in Rust.
  - [x] Entry at `rust/chains/tw_<coinid>/src/entry.rs`.
  - [x] Address at `rust/chains/tw_<coinid>/src/address.rs`.
  - [x] Transaction \(if necessary\) at `rust/chains/tw_<coinid>/src/transaction.rs`.
  - [x] Signer at `rust/chains/tw_<coinid>/src/signer.rs`.
  - [x] Compiler at `rust/chains/tw_<coinid>/src/compiler.rs`.
  - [x] Write unit tests. Put them in the `rust/chains/tw_<coinid>/tests` directory.
  - [x] Write Rust integration tests. Put them in a subfolder of `rust/tw_any_coin/tests/chains/<coinid>`
    - [x] Transaction signing tests, at least one mainnet transaction test.
    - [x] Add stake, unstake, get rewards tests if the blockchain is PoS like.
- [x] Write C++ integration tests. Put them in a subfolder of `tests/chains/<CoinId>`.
- [x] Validate generated code in Android an iOS projects. Write integration tests for each.
- [x] Extend central derivation and validation tests: make sure the following tests are extended with the new coin: `CoinAddressDerivationTests.cpp` and
      `coin_address_derivation_test.rs`,
      `CoinAddressValidationTests.cpp`,
      `TWHRPTests.cpp`,
      `CoinAddressDerivationTests.kt`,
      `CoinAddressDerivationTests.swift`.
- [x] Upload coin icon to [trustwallet/assets](https://github.com/trustwallet/assets/#how-to-add-asset) if necessary.

### CODE REVIEW

**Comment by b00f** [comment]
@philiparthurmoore 

I hope this message finds you well!

If you have a moment, I would greatly appreciate your review of this pull request. Please let us know if there’s anything we might have missed that needs attention before we proceed with the merge.

Thank you very much for your assistance!

**Comment by Milerius** [comment]
Hey @b00f He will review the PR next week - thank you

File: rust/chains/tw_pactus/src/transaction.rs
Comment (satoshiotomakan) [inline rust/chains/tw_pactus/src/transaction.rs:19]: How about adding `Copy` derive and `repr(u8)` here?

File: rust/chains/tw_pactus/src/transaction.rs
Comment (satoshiotomakan) [inline rust/chains/tw_pactus/src/transaction.rs:66]: Could you please move `TransferPayload` to a separate module?
For example,
```
/transaction/
/transaction/mod.rs
/transaction/payload/
/transaction/payload/mod.rs
/transaction/payload/transfer.rs
```

File: rust/chains/tw_pactus/src/transaction.rs
Comment (satoshiotomakan) [inline rust/chains/tw_pactus/src/transaction.rs:145]: Could you please move `Transaction::from_proto` implementation to a separate module.
We try to separate TrustWallet's Protobuf interface and Rust entities.
For example, take a look at `TxBuilder` for BNB chain:
https://github.com/trustwallet/wallet-core/blob/406abe4b0649a74279b2e7b77a2b10d1bb61520c/rust/chains/tw_binance/src/modules/tx_builder.rs#L12-L29

File: rust/chains/tw_pactus/src/transaction.rs
Comment (satoshiotomakan) [inline rust/chains/tw_pactus/src/transaction.rs:162]: Could you please remove `Bond` transaction payload if it's not supported yet? This is to avoid confusion for WalletCore users

File: rust/chains/tw_pactus/src/transaction.rs
Comment (satoshiotomakan) [inline rust/chains/tw_pactus/src/transaction.rs:287]: Can we remove other `PyalodType` enum variants?

File: rust/tw_tests/tests/chains/pactus/pactus_sign.rs
Comment (satoshiotomakan) [inline rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: Could you please add a comment like `Successfully broadcasted: https://...`? It will make it easier to find the transaction on explorer

File: tests/chains/Pactus/SignerTests.cpp
Comment (satoshiotomakan) [inline tests/chains/Pactus/SignerTests.cpp:31]: Please add a comment here as well

File: tests/chains/Pactus/TestCases.h
Comment (satoshiotomakan) [inline tests/chains/Pactus/TestCases.h:41]: These values aren't used in `SignerTests.cpp` and `CompilerTests.cpp`

File: rust/chains/tw_pactus/src/address.rs
Comment (b00f) [reply rust/chains/tw_pactus/src/address.rs]: It is resolved, please check...

File: rust/chains/tw_pactus/src/encoder/var_int.rs
Comment (b00f) [reply rust/chains/tw_pactus/src/encoder/var_int.rs:14]: It's more compact than Bitcoin.
For a better comparison, I updated the UTXO tests. In the same test, `CompactInteger` resulted in 27 bytes, while `VarInt` resulted in 21 bytes.

File: rust/chains/tw_pactus/src/transaction.rs
Comment (b00f) [reply rust/chains/tw_pactus/src/transaction.rs]: It's done, thanks.

File: rust/chains/tw_pactus/src/transaction.rs
Comment (b00f) [reply rust/chains/tw_pactus/src/transaction.rs]: It is done. Thanks for the recommendation.

File: rust/chains/tw_pactus/src/transaction.rs
Comment (b00f) [reply rust/chains/tw_pactus/src/transaction.rs]: It is really good idea to separate them. 
I can move it to `transaction_util.rs` if you want.

File: rust/tw_tests/tests/chains/pactus/pactus_sign.rs
Comment (b00f) [reply rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: Addresses in Testnet start with `tpc1`, unlike Mainnet addresses which start with `pc1`. 
And Ed25519 is not yet active on Mainnet, but it is active on Testnet. 
Can we add this test at a later time?

File: tests/chains/Pactus/TestCases.h
Comment (b00f) [reply tests/chains/Pactus/TestCases.h]: Opps, It was really bad of me. They supposed to be used! Now it is fixed. Please double check.

File: rust/chains/tw_pactus/src/transaction.rs
Comment (b00f) [reply rust/chains/tw_pactus/src/transaction.rs]: To maintain compatibility with the core code, I suggest keeping it, as it can later be compared with the source code.

File: rust/chains/tw_pactus/src/transaction.rs
Comment (b00f) [reply rust/chains/tw_pactus/src/transaction.rs]: I fully understand your concern, but I would like to request that we keep this code. Implementing all the functionality in a single PR by one author might not be ideal. I hope that soon another team member will update the code and add the bond transaction. Until then, I hope we can retain this code.

**Comment by b00f** [comment]
@satoshiotomakan Thank you for taking the time to review this PR. I believe all of your comments have been addressed. Could you please double-check and resolve those that are confirmed?

**Review by satoshiotomakan — DISMISSED** [review]
LGTM!

File: rust/tw_tests/tests/chains/pactus/pactus_sign.rs
Comment (satoshiotomakan) [reply rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: it's ok to add a comment with a testnet transaction link. Did you broadcast the transaction from the test, right? 🙂

Comment (b00f) [reply rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: I tried to do it, but the issue is that the address type needs to support Testnet, which seems unnecessary to me in this case. We will definitely add a test with a link to explorer soon. I hope this won't block us from merging the PR.

Comment (satoshiotomakan) [reply rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: One of the main criteria is to have a unit test with a successfully broadcasted mainnet transaction. It's ok if the blockchain is still testnet yet to have a testnet transaction, of if the test checks the functionality on testnet only.
When Pactus ed25519 will be on mainnet?

Comment (Ja7ad) [reply rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: > One of the main criteria is to have a unit test with a successfully broadcasted mainnet transaction. It's ok if the blockchain is still testnet yet to have a testnet transaction, of if the test checks the functionality on testnet only. But in this case, there is only one transaction example that never has been broadcasted. When Pactus ed25519 will be on mainnet?

Block 2,320,000 = https://pacviewer.com/block/2320000

Comment (b00f) [reply rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: It is scheduled to get active at first of November. Once it activated, we will add a real test here.

Comment (satoshiotomakan) [reply rust/tw_tests/tests/chains/pactus/pactus_sign.rs:40]: Hi @b00f, let's wait for the real test case with a link to the successfully broadcasted transaction.
Please note everything that WalletCore guarantees that everything in master is ready by everyone. If there are no real tests, I can't merge it.

**Comment by satoshiotomakan** [comment]
@b00f please also fix the following assertion o iOS:
```
testAddress, XCTAssertEqual failed: ("95794161374b22c696dabb98e93f6ca9300b22f3b904921fbf560bb72145f4fa") is not equal to ("0x95794161374b22c696dabb98e93f6ca9300b22f3b904921fbf560bb72145f4fa")
```

**Comment by b00f** [comment]
@satoshiotomakan it's done.

**Comment by b00f** [comment]
@satoshiotomakan 
The tests in Rust are now compatible with mainnet transactions. Additionally, we have implemented the Bond Payload, which is also covered by tests.

**Comment by b00f** [comment]
All tests have been updated with data from broadcasted transactions on the mainnet. Transactions are available here:

- An example of [Transfer](https://pacviewer.com/transaction/1b6b7226f7935a15f05371d1a1fefead585a89704ce464b7cc1d453d299d235f)
- An example of [Bond with Public Key](https://pacviewer.com/transaction/d194b445642a04ec78ced4448696e50b733f2f0b517a23871882c0eefaf1c28f)
- An example of [Bond without Public Key](https://pacviewer.com/transaction/f83f583a5c40adf93a90ea536a7e4b467d30ca4f308d5da52624d80c42adec80)

@satoshiotomakan please check. Thanks

**Review by satoshiotomakan — APPROVED** [review]
LGTM

**Comment by b00f** [comment]
@satoshiotomakan There was a typo in the Android test, which has been fixed. Please re-run the tests. Apologies for the mistake.

**Review by satoshiotomakan — APPROVED** [review]

**Review by Milerius — APPROVED** [review]

**Comment by satoshiotomakan** [comment]
Hi @b00f, if you want the Pactus Blockchain to be integrated into TrustWallet app, please submit a request to
https://support.trustwallet.com/en/support/tickets/new

**Comment by b00f** [comment]
@SatoshiOtomakan and the Wallet-Core project team,

Thank you for this amazing project and for carefully reviewing this PR and helping us correct our mistakes. I will definitely submit the request, and we hope to see Pactus in the Trust Wallet application soon.

**Pull request merged**
